deeplake/core/vectorstore/vector_search/dataset/dataset.py

```diff
@@ -385,6 +389,65 @@ def update_embedding_info(logger, dataset, embedding_function):
     set_embedding_info(dataset[embeddings_tensors[0]], embedding_function)
 
 
+def extend(
+    embedding_function: List[Callable],
+    embedding_data: List[Any],
+    embedding_tensor: Union[str, List[str]],
+    processed_tensors: Dict[str, List[Any]],
+    dataset: deeplake.core.dataset.Dataset,
+    max_bytes_per_minute: int = MAX_BYTES_PER_MINUTE,
+    target_byte_size: int = TARGET_BYTE_SIZE,
+):
+    """
+    Function to extend the dataset with new data.
+
+    Args:
+        embedding_function (List[Callable]): List of embedding functions to be used to create embedding data.
+        embedding_data (List[Any]): List of data to be embedded.
+        embedding_tensor (Union[str, List[str]]): Name of the tensor(s) to store the embedding data.
+        processed_tensors (Dict[str, List[Any]]): Dictionary of tensors to be added to the dataset.
+        dataset (deeplake.core.dataset.Dataset): Dataset to be extended.
+
+
+    """
+    if embedding_function:
+        for func, data, tensor in zip(
+            embedding_function, embedding_data, embedding_tensor
+        ):
+            data_batched = chunk_by_bytes(data, target_byte_size=target_byte_size)
+
+            # Calculate the number of batches you can send each minute
+            batches_per_minute = max_bytes_per_minute / target_byte_size
+
+            # Calculate sleep time in seconds between batches
+            sleep_time = 60 / batches_per_minute
+
+            embedded_data = []
+
+            for data_i in tqdm(
+                data_batched, total=len(data_batched), desc="Creating embedding data"
+            ):
+                start = time.time()
+                embedded_data.append(func(data_i))
+                end = time.time()
+                if func.__module__ == "langchain.embeddings.openai":
```

Comment:
Is this too tightly coupled to openai embeddings? Other rate-limited APIs would have similar needs, and we don't necessarily know them all. And/or limits may change over time?

I like how it automatically works right for openai, but maybe instead of the arguments always having a value and sleeping purely on the module name we have the argument value optional. 
Then:
- If set to a value we always sleep regardless of the module name
- If set to 0 we never sleep regardless of the module name
- If not set, we do what we think is right (your logic of sleeping a constant-based amount based on the module name)

?

Reply:
interesting suggestions, but that will also affect our api right? We might need to add additional variable to vectorstore api and also LangChain one right? It is definately worth exploring but, the thing is that most of the people even doesn't know what their rate limit is. Currently openai is the only platform that is raising this exception. So as an initial solution, I am thinking to ship this to improve user experience with vectorstore + openai and later add more granularity to the way we manage rate limits (Like changing api and so on)

Reply:
Yes, it would need to be exposed/defined at the user-facing API level, including the vectorstore API.

It's fine to have an initial release that doesn't expose it but have a plan for it, but want to make sure it won't be a breaking change to introduce. Like the new `extend` API you added here has  `max_bytes_per_minute` and `target_byte_size` arguments which users will now see and be able to use, right? So that will be forcing us to work around those particular arguments with the fact that users may be relying on a default value being set down the road?

If we want the flexibility later on to define a different way of controlling the behavior, maybe don't introduce those arguments at all now and just use the constants in the function?

Reply:
ok, let me remove these args as they don't serve any particular purpose